### PR TITLE
fix(settlement): fixed submission bug where multiple events would cancel subscription 

### PR DESCRIPTION
**Pull request by omritoptix**
This PR solves few issues:
1. increased buffer for hub events
2. filtering events
3. Allow for retries on polling when waiting for batch inclusion

# PR Standards

## Opening a pull request should be able to meet the following requirements

---

Close #XXX

<-- Briefly describe the content of this pull request -->

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

**Comment by danwt** [comment]
@omritoptix there is a conflict
is it ready for review afterwards?

**Comment by omritoptix** [comment]
@danwt , @mtsitrin is the owner of this now. I think we should resolve the conflict.

**Comment by mtsitrin** [comment]
ready for review

File: settlement/dymension/dymension.go [hunk 1]
```diff
@@ -212,76 +212,96 @@ func (d *HubClient) PostBatch(batch *types.Batch, daClient da.Client, daResult *
...
-					batch.StartHeight,
-					"endHeight",
-					batch.EndHeight,
-					"error",
-					err,
-				)
-
-				timer.Stop() // we don't forget to clean up
-				continue
+		batchAcceptanceAttempts := 5
```
Comment (omritoptix) [inline settlement/dymension/dymension.go:255]: think it's cleaner to have the `5` defined as a global const.

File: settlement/dymension/dymension.go [hunk 2]
```diff
@@ -212,76 +212,96 @@ func (d *HubClient) PostBatch(batch *types.Batch, daClient da.Client, daResult *
...
+				// Check if the batch was accepted by the settlement layer, and we've just missed the event.
+				includedBatch, err := d.pollForBatchInclusion(batch.EndHeight)
+				if err == nil && !includedBatch {
+					// no error, but still not included
+					batchAcceptanceAttempts--
+					if batchAcceptanceAttempts > 0 {
+						timer.Reset(d.batchAcceptanceTimeout)
+						continue
+					}
+					err = fmt.Errorf("batch not accepted by the settlement layer")
```
Comment (omritoptix) [inline settlement/dymension/dymension.go:284]:
```suggestion
					err = fmt.Errorf("timed out waiting for batch inclusion on settlement layer")
```

File: settlement/dymension/dymension.go (hunk 1 above)
Comment (mtsitrin) [reply settlement/dymension/dymension.go:255]: i'll add it in this PR
https://github.com/dymensionxyz/dymint/pull/863

File: settlement/dymension/dymension.go [hunk 3]
```diff
@@ -215,76 +216,96 @@ func (d *HubClient) PostBatch(batch *types.Batch, daClient da.Client, daResult *
...
-					batch.EndHeight,
-					"error",
-					err,
-				)
-
-				timer.Stop() // we don't forget to clean up
-				continue
+		batchAcceptanceAttempts := d.batchAcceptanceAttempts
+
+		for {
```
Comment (danwt) [inline settlement/dymension/dymension.go:261]: pretty sure this top level for never has effect?

File: settlement/dymension/dymension.go [hunk 4]
```diff
@@ -215,76 +216,96 @@ func (d *HubClient) PostBatch(batch *types.Batch, daClient da.Client, daResult *
...
+				d.logger.Info("Batch accepted", "startHeight", batch.StartHeight, "endHeight", batch.EndHeight)
+				return nil
 			}
-
-			// all good
-			d.logger.Info("Batch accepted", "startHeight", includedBatch.StartHeight, "endHeight", includedBatch.EndHeight)
-			return nil
 		}
 	}
 }
```
Comment (danwt) [inline settlement/dymension/dymension.go:311]: this whole block is getting pretty complicated, can we extract some methods (`submitBatch`, `waitForInclusion`) and try to simplify (especially to simplify the submission part)

File: settlement/dymension/dymension.go [hunk 5]
```diff
@@ -400,10 +421,13 @@ func (d *HubClient) submitBatch(msgUpdateState *rollapptypes.MsgUpdateState) err
 
 func (d *HubClient) eventHandler() {
 	// TODO(omritoptix): eventsChannel should be a generic channel which is later filtered by the event type.
-	eventsChannel, err := d.client.SubscribeToEvents(d.ctx, "dymension-client", fmt.Sprintf(eventStateUpdate, d.config.RollappID))
+	subscriber := fmt.Sprintf("dymension-client-%s", uuid.New().String())
+	eventsChannel, err := d.client.SubscribeToEvents(d.ctx, subscriber, fmt.Sprintf(eventStateUpdate, d.config.RollappID), 100)
```
Comment (danwt) [inline settlement/dymension/dymension.go:425]: ditto discussion on 100 as a magic value from previous comment

**Review by danwt — CHANGES_REQUESTED** [review]
I do think it would be good to simplify the big loop

File: settlement/dymension/dymension.go (hunk 3 above)
Comment (mtsitrin) [reply settlement/dymension/dymension.go:261]: it should be called in case of error in the waitForInclusion loop

Comment (danwt) [reply settlement/dymension/dymension.go:261]: https://github.com/dymensionxyz/dymint/blob/f76f0bcff1cd14c0e70d1b317479d51717ba27c7/settlement/dymension/dymension.go#L262

defo need to factor out and make sure this loop is simple and works

File: settlement/dymension/dymension.go (hunk 4 above)
Comment (mtsitrin) [reply settlement/dymension/dymension.go:311]: simplified a bit

**Review by danwt — APPROVED** [review]
nice!

**Review by srene — APPROVED** [review]

**Pull request merged**